### ISSUE TITLE
GitRepo.get_hexsha() can be extremely slow (6min)

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4801]
TL;DR: Should we switch from `git-show` to `git-log` for `GitRepo.format_commit()`. Or is there a way to further limit processing in `git-show` beyond `--no-patch`.

I discovered this via a `datalad status` call, which I thought would be hanging.

<details>
<summary>Here is the log level 1 output</summary>

```
% dl -l 1 status -e commit --untracked no -t raw
[DEBUG  ] Command line args 1st pass. Parsed: Namespace() Unparsed: ['status', '-e', 'commit', '--untracked', 'no', '-t', 'raw']                                                          [DEBUG  ] Discovering plugins                                                                [Level 5] Importing module datalad.core.local.status
[DEBUG  ] Building doc for <class 'datalad.core.local.status.Status'>                        [Level 5] Finished setup_parser                   
[DEBUG  ] Parsing known args among ['/usr/bin/datalad', '-l', '1', 'status', '-e', 'commit',
'--untracked', 'no', '-t', 'raw']                                      
[Level 5] Running: ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin']             
[Level 8] Finished running ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin'] with
status 0                                                                                     
[Level 2] Entered eval_func for <function Status.__call__ at 0x7f119481f378>                
[DEBUG  ] Determined class of decorated function: <class 'datalad.core.local.status.Status'>
[Level 2] Returning generator_func from eval_func for <class 'datalad.core.local.status.Statu
s'>
[Level 5] Parsed ri /data/project/enki/processed into fields {'path': '/data/project/enki/pro
cessed', 'scheme': '', 'username': '', 'password': '', 'hostname': '', 'port': '', 'query': '
', 'fragment': ''}
[Level 5] Detected file ri
[DEBUG  ] Resolved dataset for status reporting: /data/project/enki/processed
[Level 5] Detected <class 'datalad.support.annexrepo.AnnexRepo'> at /data/project/enki/processed
[Level 5] Running: ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin']
[Level 8] Finished running ['git', '--git-dir=', 'config', '-z', '-l', '--show-origin'] with status 0
[Level 5] Found 1 previous connections
[Level 5] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/data/project/enki/processed/.datalad/config']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin', '--file', '/data/project/enki/processed/.datalad/config'] with status 0
[Level 5] Running: ['git', 'config', '-z', '-l', '--show-origin']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin'] with status 0
[Level 5] Parsed ri ssh://bulk1.htc.inm7.de/ds/enki/srv/669/7c1b8-b617-11ea-b713-b4969157768c
 into fields {'path': '/ds/enki/srv/669/7c1b8-b617-11ea-b713-b4969157768c', 'scheme': 'ssh',
'username': '', 'password': '', 'hostname': 'bulk1.htc.inm7.de', 'port': '', 'query': '', 'fr
agment': ''}
[Level 5] Detected url ri                                                           [22/9047]
[Level 5] Parsed ri git@jugit.fz-juelich.de:inm7/datasets/enki-processed.git into fields {'pa
th': 'git@jugit.fz-juelich.de:inm7/datasets/enki-processed.git', 'scheme': '', 'username': ''
, 'password': '', 'hostname': '', 'port': '', 'query': '', 'fragment': ''}
[Level 5] Detected ssh ri
[DEBUG  ] query AnnexRepo(/data/project/enki/processed).diffstatus() for paths: None
[Level 5] Running: ['git', 'show', '-z', '--no-patch', '--format=%H', '--']
[Level 8] Finished running ['git', 'show', '-z', '--no-patch', '--format=%H', '--'] with status 0
[DEBUG  ] AnnexRepo(/data/project/enki/processed).get_content_info(...)
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard']
[Level 5] Running: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard']
[Level 8] Finished running ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] with status 0
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard']
[DEBUG  ] Done AnnexRepo(/data/project/enki/processed).get_content_info(...)
[Level 5] Running: ['git', 'ls-files', '-z', '-m']
[Level 8] Finished running ['git', 'ls-files', '-z', '-m'] with status 0
[DEBUG  ] AnnexRepo(/data/project/enki/processed).get_content_info(...)
[DEBUG  ] Query repo: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l']
[Level 5] Running: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l']
[Level 8] Finished running ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l'] with status 0
[DEBUG  ] Done query repo: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l']
[DEBUG  ] Done AnnexRepo(/data/project/enki/processed).get_content_info(...)
[Level 5] Parsed ri /data/project/enki/processed/code/pipelines into fields {'path': '/data/p
roject/enki/processed/code/pipelines', 'scheme': '', 'username': '', 'password': '', 'hostnam
e': '', 'port': '', 'query': '', 'fragment': ''}
[Level 5] Detected file ri
[Level 5] Running: ['git', 'show', '-z', '--no-patch', '--format=%H', '--']
[Level 8] Finished running ['git', 'show', '-z', '--no-patch', '--format=%H', '--'] with status 0
[Level 5] Parsed ri /data/project/enki/processed/fmriprep into fields {'path': '/data/project
/enki/processed/fmriprep', 'scheme': '', 'username': '', 'password': '', 'hostname': '', 'por
t': '', 'query': '', 'fragment': ''}
[Level 5] Detected file ri
[Level 5] Running: ['git', 'show', '-z', '--no-patch', '--format=%H', '--']
[Level 8] Finished running ['git', 'show', '-z', '--no-patch', '--format=%H', '--'] with status 0
[Level 5] Parsed ri /data/project/enki/processed/freesurfer into fields {'path': '/data/proje
ct/enki/processed/freesurfer', 'scheme': '', 'username': '', 'password': '', 'hostname': '',
'port': '', 'query': '', 'fragment': ''}
[Level 5] Detected file ri
[Level 5] Running: ['git', 'show', '-z', '--no-patch', '--format=%H', '--']

... and now it hangs
```

</details>

Essentially `[Level 5] Running: ['git', 'show', '-z', '--no-patch', '--format=%H', '--'] ` does not seem to return.

<details>
<summary>Traceback when killed while hanging (runner is waiting for the process to end, it seems)</summary>

```
Interrupted by user while doing magic: KeyboardInterrupt() [selectors.py:select:415] [2/9657]
Traceback (most recent call last):            
  File "/usr/bin/datalad", line 8, in <module>                                              
    main()
  File "/usr/lib/python3/dist-packages/datalad/cmdline/main.py", line 521, in main
    ret = cmdlineargs.func(cmdlineargs)
  File "/usr/lib/python3/dist-packages/datalad/interface/base.py", line 786, in call_from_par
ser
    ret = list(ret)
  File "/usr/lib/python3/dist-packages/datalad/interface/utils.py", line 413, in generator_fu
nc
    allkwargs):
  File "/usr/lib/python3/dist-packages/datalad/interface/utils.py", line 552, in _process_res
ults
    for res in results:
  File "/usr/lib/python3/dist-packages/datalad/core/local/status.py", line 419, in __call__
    content_info_cache):
  File "/usr/lib/python3/dist-packages/datalad/core/local/status.py", line 121, in _yield_sta
tus
    _cache=cache)
  File "/usr/lib/python3/dist-packages/datalad/support/gitrepo.py", line 3736, in diffstatus
    subrepo_commit = subrepo.get_hexsha()
  File "/usr/lib/python3/dist-packages/datalad/support/gitrepo.py", line 1747, in get_hexsha
    commitish)
  File "/usr/lib/python3/dist-packages/datalad/support/gitrepo.py", line 1718, in format_comm
it
    '', cmd, expect_stderr=True, expect_fail=True)
  File "/usr/lib/python3/dist-packages/datalad/support/gitrepo.py", line 328, in newfunc
    result = func(self, files_new, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/datalad/support/gitrepo.py", line 2058, in _git_custom
_command
    expect_fail=expect_fail)
  File "/usr/lib/python3/dist-packages/datalad/cmd.py", line 143, in run_gitcommand_on_file_l
ist_chunks
    results.append(func(cmd, *args, **kwargs))
  File "/usr/lib/python3/dist-packages/datalad/cmd.py", line 1141, in run
    *args, **kwargs)
  File "/usr/lib/python3/dist-packages/datalad/cmd.py", line 982, in run
    raise exc_info[1]
  File "/usr/lib/python3/dist-packages/datalad/cmd.py", line 930, in run
    out = proc.communicate()
  File "/usr/lib/python3.7/subprocess.py", line 939, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.7/subprocess.py", line 1681, in _communicate
    ready = selector.select(timeout)
  File "/usr/lib/python3.7/selectors.py", line 415, in select
    fd_event_list = self._selector.poll(timeout)
KeyboardInterrupt

```

</details>

`strace` of the `git-show` process is not informative (nothing seems to be going on)

```
% strace -p 8068
strace: Process 8068 attached
strace: [ Process PID=8068 runs in x32 mode. ]                                              
```

but htop reports that process to be running at 100% CPU with 14.5G resident memory size (for 5min now).

```
datalad 0.13.1
git version 2.24.0
git-annex version: 8.20200330 | 7.20190819+git2-g908476a9b-1~ndall+1
```

Update: It did not actually hang, but ran for 6min. I think that it might not be running in the repo that I thought it should run in (we might want to add a runner's CWD to the logging output).

Update2: Indeed, the CWD of git-show point to a subdataset, and as visible in the traceback above this call is triggered by a `GitRepo.get_hexsha()` call.

Now in this particular call it is asking for the hexsha of HEAD (`commitish=None`). I can get this in less than 10ms vs 6min with `git-show`:

```
% time git log -1 --format=%H | cat
4c3df804a89bd2417e0d51dddb38a6129c1993c5
git log -1 --format=%H  0.00s user 0.00s system 83% cpu 0.009 total
```

The nature of the last commit is a giant octopus-merge of 1000 branches. it seems that `git-show` is doing a lot more inspection. I have not (yet) determined a way to have `git-show` do less.

**Comment by yarikoptic** [comment]
lovely!  but I wonder if there is no other command even "lower" level than `log`...

```
$> time git log -1 --format=%H           
119b9305772e6e5de191667c60c8a6ad483e4b60
git log -1 --format=%H  0.08s user 0.02s system 98% cpu 0.103 total

$> time git show-ref refs/`cat .git/HEAD`
119b9305772e6e5de191667c60c8a6ad483e4b60 refs/heads/enh-config-parallel
git show-ref refs/`cat .git/HEAD`  0.02s user 0.03s system 98% cpu 0.043 total
```
I win!

**Comment by kyleam** [comment]
@mih
> it seems that `git-show` is doing a lot more inspection.

That's good to know.  I wouldn't have guessed that.

For `format_commit`, where we need to support a format argument, I don't see any issue with switching to `git log -1 ...`.

Were I writing `get_hexsha`, I'd say `rev-parse [--short] X^{commit}`.  I also think `show-ref -s --verify` (@yarikoptic's suggestion) would be fine and don't expect there'd be much difference in speed.

@yarikoptic
>     git show-ref refs/`cat .git/HEAD`

When on a branch, .git/HEAD starts with "ref:".  That's matching because by default show-ref takes a pattern or patterns.  `--verify` requires an exact match:

```
$ git show-ref refs/heads/master
90e218937592458c9bdec9d97a10a68c4772e14f refs/heads/master
6c5e8e6b5c60866325b40a12638ae373d34b9d73 refs/wip/index/refs/heads/master
1d9a66e311adbd2d630e858933f8c57deae33cba refs/wip/wtree/refs/heads/master

$ git show-ref --verify refs/heads/master
90e218937592458c9bdec9d97a10a68c4772e14f refs/heads/master
```
